Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp

```diff
@@ -0,0 +1,148 @@
+                  "Integer that initializes a random-number generator, good for reproducibility");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("OutputWorkspace", "", Direction::Output),
+                  "Name of output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         std::mt19937 &gen) {
+  Mantid::HistogramData::HistogramY outputY(cdf.size(), 0.0);
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+  for (int i = 0; i < numIterations; ++i) {
+    double randomNum = dis(gen);
+    // Find the bin corresponding to the random number in the CDF
+    auto it = std::lower_bound(cdf.begin(), cdf.end(), randomNum);
+    size_t index = std::distance(cdf.begin(), it);
+
+    // Ensure the index is within bounds
+    if (index < outputY.size()) {
+      outputY[index] += 1.0;
+    }
+  }
+  return outputY;
+}
+
+std::vector<double> CreateMonteCarloWorkspace::computeNormalizedCDF(const Mantid::HistogramData::HistogramY &yData) {
+  std::vector<double> cdf(yData.size());
+  std::partial_sum(yData.begin(), yData.end(), cdf.begin());
+  double total_counts = cdf.back();
+  // Normalize the CDF
+  std::transform(cdf.begin(), cdf.end(), cdf.begin(), [total_counts](double val) { return val / total_counts; });
+  return cdf;
+}
+
+int CreateMonteCarloWorkspace::computeNumberOfIterations(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
+  return static_cast<int>(std::round(total_counts));
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+
+// Using Cumulative Distribution Function
+void CreateMonteCarloWorkspace::exec() {
+  // Progress Bar set up
+  int numSteps = 7;
+  API::Progress progress(this, 0.0, 1.0, numSteps);
+
+  MatrixWorkspace_sptr instWs = getProperty("InputWorkspace");
+  int seed_input = getProperty("Seed");
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  const Mantid::HistogramData::HistogramY &yData = instWs->y(0); // Counts in each bin
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  int numIterations = computeNumberOfIterations(yData);
+  std::vector<double> cdf = computeNormalizedCDF(yData);
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  MatrixWorkspace_sptr outputWs = WorkspaceFactory::Instance().create(instWs);
+  progress.report();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Maybe a stupid question, sorry, but why do you need these sleep statements?

Reply:
I think you might've have the progress bar a little wrong (although I'm no expert).

If your algorithm, for example, iterated over the spectra of an input workspace and performed some kind of operation with each, you would have the progress bar update after each spectra operation. I think that's the idea anyway, it's showing the progress of the main bulk of the computation. I think for you this would be in `fillHistogramWithRandomData`.

I think you can also set it to display text (e.g. "writing output"). Maybe you can have a look at some other algorithms and see how it's used there.

Reply:
I updated the code for the progress bar, although personally I prefer the previous look, the new version does not need the sleep statements.